hpx/parallel/algorithms/merge.hpp

```diff
@@ -249,23 +249,38 @@ namespace hpx { namespace parallel { inline namespace v1
         template <typename ExPolicy,
             typename RandIter1, typename RandIter2, typename RandIter3,
             typename Comp, typename Proj1, typename Proj2>
-        hpx::util::tuple<RandIter1, RandIter2, RandIter3>
+        hpx::future<hpx::util::tuple<RandIter1, RandIter2, RandIter3>>
         parallel_merge(ExPolicy && policy,
             RandIter1 first1, RandIter1 last1,
             RandIter2 first2, RandIter2 last2,
             RandIter3 dest, Comp && comp,
             Proj1 && proj1, Proj2 && proj2)
         {
-            parallel_merge_helper(std::forward<ExPolicy>(policy),
-                first1, last1, first2, last2, dest,
-                std::forward<Comp>(comp),
-                std::forward<Proj1>(proj1),
-                std::forward<Proj2>(proj2),
-                false,
-                lower_bound_helper());
+            typedef hpx::util::tuple<RandIter1, RandIter2, RandIter3>
+                result_type;
+
+            hpx::future<result_type> f = execution::async_execute(
+                policy.executor(),
+                [=]() mutable -> result_type
+                {
+                    try {
+                        parallel_merge_helper(
+                            policy, first1, last1, first2, last2, dest,
+                            comp, proj1, proj2, false, lower_bound_helper());
+
+                        return hpx::util::make_tuple(last1, last2,
+                            dest + (last1 - first1) + (last2 - first2));
+                    }
+                    catch (...) {
+                        util::detail::handle_local_exceptions<ExPolicy>::call(
+                            std::current_exception());
+                    }
 
-            return hpx::util::make_tuple(last1, last2,
-                dest + (last1 - first1) + (last2 - first2));
+                    // Not reachable.
+                    HPX_ASSERT(false);
+                });
```

Comment:
Could we prefect forward `comp`, `proj1` and `proj2` into the lambda?

Reply:
@sithhell I did. There was 3 key points.
1. The lambda must have own variables including ```comp```, ```proj1```, and ```proj2``` because the variables will be used in asynchronous manner.
2. So, the lambda must construct own variables from original parameters. In this point, the parameters must be perfect forwarded into constructors of the own variables.
3. Support C++11.

Reply:
@sithhell How do you think about my modification for perfect forwarding?
There are some places that this way can be applied to. (For example, https://github.com/STEllAR-GROUP/hpx/blob/a5ac8fc04507d719414f14431296148850adac93/hpx/parallel/util/scan_partitioner.hpp#L342-L350)

If this way is appropriate enough, I will apply it to some places, too.